### PR TITLE
remove impersonation restriction

### PR DESCRIPTION
**Pull request by KHOIOHK87**
Related to this issue: https://github.com/razee-io/razeedeploy-core/issues/189

Admission webhook was created [HERE](https://github.com/razee-io/ImpersonationWebhook) to validate if an authenticated user has impersonation permission. The restriction in the base controllers should be removed since the webhook already handles validation. This means that razee controllers and the webhook must be deployed together.

**Potential breaking changes:**
- Currently, anyone can impersonate others when creating resources in `razeedeploy` namespace. With this change, this will not be allowed if authenticated users are not granted impersonate permission, and the `impersonateUser` field will be set to the authenticated user by the webhook.
- Currently, `impersonateUser` field is set to `razeedeploy` by default if authenticated user tries to impersonate others when creating resources outside of `razeedeploy` namespace. With this change, the default value for `impersonateUser` field will be the authenticated user.

### CODE REVIEW

File: lib/BaseController.js [hunk 1]
```diff
@@ -85,16 +85,8 @@ module.exports = class BaseController {
...
-      if (impersonateUser !== 'razeedeploy') {
-        if (impersonateUser !== 'false' && this.namespace === 'razeedeploy') {
-          this._kubeResourceMeta.addHeader('Impersonate-User', impersonateUser);
-        } else {
-          impersonateUser = 'razeedeploy';
-          return await this.patchSelf({ 'spec': { 'clusterAuth': { 'impersonateUser': impersonateUser } } });
-        }
-      }
+      let impersonateUser = objectPath.get(this._data, 'object.spec.clusterAuth.impersonateUser', 'razeedeploy');
+      this._kubeResourceMeta.addHeader('Impersonate-User', impersonateUser);
```
Comment (alewitt2) [inline lib/BaseController.js:89]: if impersonateUser is razeedeploy, we **shouldn't** add the header.

File: lib/CompositeController.js [hunk 2]
```diff
@@ -141,13 +141,8 @@ module.exports = class CompositeController extends BaseController {
...
     }
 
-    // TODO: https://github.com/razee-io/razeedeploy-core/issues/189
     let impersonateUser = objectPath.get(this.data, 'object.spec.clusterAuth.impersonateUser', 'razeedeploy');
-    if (impersonateUser !== 'razeedeploy' && this.namespace === 'razeedeploy') {
-      krm.addHeader('Impersonate-User', impersonateUser);
-    } else {
-      impersonateUser = 'razeedeploy';
-    }
+    krm.addHeader('Impersonate-User', impersonateUser);
```
Comment (alewitt2) [inline lib/CompositeController.js:145]: if impersonateUser is razeedeploy, we **shouldn't** add the header.

File: lib/BaseController.js (hunk 1 above)
Comment (KHOIOHK87) [reply lib/BaseController.js]: addressed.

File: lib/CompositeController.js (hunk 2 above)
Comment (KHOIOHK87) [reply lib/CompositeController.js]: addressed.

**Comment by KHOIOHK87** [comment]
> **Breaking changes:**
> 
> * anyone who is currently using razee wont have the webhook deployed, and therefor all user will be able to do impersonation in any namespace once updated. It should be heavily stressed that users should deploy the webhook before upgrading version to prevent unauthorized users being allowed to do this.

Agreed. We can do this in release notes. I also updated the installation instruction for applicable controllers to install the webhook first, before install the controllers. Razee deploy delta also updated. Added you to those PRs.

**Review by alewitt2 — CHANGES_REQUESTED** [review]
as discussed, we should add a configuration option to `razeedeploy-config` configmap to enable this feature for the whole cluster instead of automatically enabling it for the whole cluster. if not specified, default to disabled.

File: lib/BaseController.js [hunk 3]
```diff
@@ -125,6 +117,33 @@ module.exports = class BaseController {
     }
   }
 
+  async preprocessImpersonation() {
+    let impersonationEnabled = await this._impersonation_enabled();
+    let impersonateUser = objectPath.get(this.data, 'object.spec.clusterAuth.impersonateUser', 'false');
+    if (impersonateUser === 'false' || (!impersonationEnabled && impersonateUser != 'razeedeploy' && this.namespace != 'razeedeploy')) {
```
Comment (alewitt2) [inline lib/BaseController.js:123]: writing down for my own thoughts:
- `if (impersonateUser === 'false')`: if this statement is true, it means the webhook is disabled and didnt enforce the resource to have a `clusterAuth.impersonateUser`. in that case we want to patch the resource to make it clear we are running all processing as `razeedeploy`
- `if (!impersonationEnabled && impersonateUser != 'razeedeploy' && this.namespace != 'razeedeploy'))`: if this statement is true, it means impersonation is disabled, some non `razeedeploy` user was defined, and it was deployed to a namespace other than `razeedeploy`. basically, when impersonation is disabled, we want any resource deployed to any namespace other than `razeedeploy` to run as the `razeedeploy` user.
- if impersonation is enabled, or the `razeedeploy` user was already defined, or the resource is in the razeedeploy namespace, we dont need to enforce the razeedeploy user on the resource.

File: lib/BaseController.js [hunk 4]
```diff
@@ -125,6 +117,33 @@ module.exports = class BaseController {
...
+    }
+    else {
+      if (this.namespace === 'razeedeploy') {
+        krm.addHeader('Impersonate-User', impersonateUser);
+      }
+    }
+
+    return impersonateUser;
+  }
+
```
Comment (alewitt2) [inline lib/BaseController.js:146]:
```suggestion
  async processImpersonation(krm) {
    let impersonateUser = objectPath.get(this.data, 'object.spec.clusterAuth.impersonateUser', 'razeedeploy');
    if (impersonateUser === 'razeedeploy') {
      return impersonateUser;
    }

    let impersonationEnabled = await this._impersonation_enabled();
    if (impersonationEnabled || this.namespace === 'razeedeploy') {
      krm.addHeader('Impersonate-User', impersonateUser);
    } else {
      impersonateUser = 'razeedeploy'
    }

    return impersonateUser;
  }

```

**Review by alewitt2 — CHANGES_REQUESTED** [review]

**Review by alewitt2 — APPROVED** [review]

**Pull request merged**